### ISSUE TITLE
Focus should be managed between markers and popups

### ISSUE THREAD

**Issue by nfreear** [nfreear/leaflet.a11y#3]
## Problem

Focus order should be logical and meaningful. At present, when a popup is opened in Leaflet, keyboard focus is not placed on the popup. When the popup is closed, focus is not placed back on the triggering marker element.

* WCAG Success Criteria [2.4.3 Focus Order](https://www.w3.org/TR/WCAG21/#focus-order).
* Core Leaflet bug(s): Leaflet/Leaflet/issues/8115, Leaflet/Leaflet/issues/8113.

## Solution

On popup open:
```js
popupElem.role = 'dialog';
popupElem.setAttribute('tabindex', '-1');
popupElem.focus();

triggerElem.ariaExpanded = 'true';
```

On popup close:
```js
if (triggerElem) {
  triggerElem.focus();
  triggerElem.ariaExpanded = 'false';
}
```

Leaflet API / plugin dev. notes:
* Handle the [`popupopen` and `popupclose` Leaflet events](https://leafletjs.com/reference.html#map-event).

---

**Comment by robinmetral** [comment]
I see that some of this has been addressed (cf. [demo](https://nfreear.github.io/leaflet.a11y)). Is it considered closed? On a related note, popups should probably close when pressing ESC, and perhaps they should also trap focus (modal dialog pattern)—thoughts?

**Comment by nfreear** [comment]
Hi @robinmetral,

Thanks for the comments. To address your points:

> Is it considered closed?
>
Yes. I need to tidy up!

> ... popups should probably close when pressing ESC.
>
Good point. I've raised this in #22.

> ... perhaps they should also trap focus (modal dialog pattern)—thoughts?
>
In the general case, pop-ups triggered when activating a map marker should not be modal, so I don't think they should trap focus.

(FYI, in Leaflet/Leaflet/issues/7968, a contributor argues that map pop-ups are not always dialogs.)

If a modal dialog was needed for a specific application, you could extend the built-in [Popup class](https://leafletjs.com/reference.html#popup).

I hope that helps.

Yours,

Nick

---
P.S. I'm always interested to know if and/or where you're thinking of using `Leaflet.a11y` plugin :wink:

You may also find [Leaflet.translate](https://github.com/nfreear/Leaflet.translate) useful.

**Comment by robinmetral** [comment]
Thanks for the detailed follow-up! Not using the plugin at the moment, still evaluating mapping libraries options for a personal project. I definitely will if going for Leaflet (plus I'd be happy to join efforts improving a11y, code and docs) :slightly_smiling_face: 

Cheers!
